Libraries/acdriver/Sources/CompileActionImageSet.cpp

```diff
@@ -0,0 +1,498 @@
+        png_set_filler(png_struct_ptr, 0xff, PNG_FILLER_AFTER);        
+    }
+
+    png_set_alpha_mode(png_struct_ptr, PNG_ALPHA_PREMULTIPLIED, PNG_DEFAULT_sRGB);
+
+    number_of_passes = png_set_interlace_handling(png_struct_ptr);
+    (void) number_of_passes;
+
+    /* Make transforms take effect */
+    png_read_update_info(png_struct_ptr, info_struct_ptr);
+
+    if (!png_get_IHDR (png_struct_ptr, info_struct_ptr,
+        &width, &height, &bit_depth, &color_type, &interlace_method, &compression_method, &filter_method)) {
+        png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+        result->normal(
+            Result::Severity::Error,
+            "png_get_IHDR returned error",
+            std::string(filename));
+        return false;
+    }
+
+    /* color_type will still show original format */
+    switch(color_type) {
+        case PNG_COLOR_TYPE_GRAY:
+            /* This is transformed to GA8 */
+        case PNG_COLOR_TYPE_GRAY_ALPHA:
+            channels = 2;
+            break;
+        case PNG_COLOR_TYPE_PALETTE:
+        case PNG_COLOR_TYPE_RGB:
+            /* These are transformed to RGBA */
+        case PNG_COLOR_TYPE_RGB_ALPHA:
+            channels = 4;
+            break;
+        default:
+            png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+            fclose(fp);
+            result->normal(
+                Result::Severity::Error,
+                "Unhandled PNG color type");
+            return false;
+    }
+
+    row_bytes = png_get_rowbytes (png_struct_ptr, info_struct_ptr);
+    if (row_bytes != (width * (bit_depth/8) * channels)) {
+        png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+        result->normal(
+            Result::Severity::Error,
+            "Unable to transform PNG pixel data");
+        return false;
+    }
+
+    row_pointers = (png_byte **) malloc (height * sizeof (png_bytep));
+    if (row_pointers == NULL)
+    {
+        png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+        result->normal(
+            Result::Severity::Error,
+            "Could not allocate memory");
+        return false;
+    }
+
+    pixels = std::vector<unsigned char>(width * height * channels);
+    unsigned char *bytes = static_cast<unsigned char *>(pixels.data());
+    for (int row = 0; row < height; row++) {
+        row_pointers[row] = bytes + (row * row_bytes);
+    }
+    png_read_image(png_struct_ptr, row_pointers);
+
+    /* clean up */
+    png_read_end (png_struct_ptr, info_struct_ptr);
+    png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, (png_infopp) NULL);
+    free(row_pointers);
+    fclose(fp);
+
+    if (width_out) {
+        *width_out = width;
+    }
+    if (height_out) {
+        *height_out = height;
+    }
+    if (channels_out) {
+        *channels_out = channels;
+    }
+    return true;
+}
+
+static uint16_t
+GenerateIdentifier(void) {
+    static uint16_t last = 0;
```

Comment:
Well, as long as MemoryFilesystem is consistent with ordering, it seems fine to me. A hashing scheme based on the name sounds fine also, but again, with only 16 bits, resolving collisions in one pass will probably also have different results based on the ordering of directory contents.